### PR TITLE
Chloeb/implement undelete

### PR DESCRIPTION
**Pull request by chloebarker**
## Overview

Implementing undelete (followup to [#1198](https://github.com/square/keywhiz/pull/1198)). Here is how it works on the table level:

Deletion:

image

Undeletion:

image

## Test case 1

Testing undeleting the secret I deleted when testing [#1198](https://github.com/square/keywhiz/pull/1198):

First make sure the secret is there:

```
> keywhiz.cli describe secret --name chloeb-test-secret-20230307-2 --include-deleted
No non-deleted secret found.
Deleted Secrets: found 1
chloeb-test-secret-20230307-2
	Internal ID:
		-5209310080942603851
	Owner:
	Created by:
		chloeb
	Created at:
		2023-03-07T20:54:50Z
	Updated by:
		chloeb
	Updated at:
		2023-03-07T20:54:50Z
```

Make sure things look right in the db:

```
> select id, name, current from secrets where id=-5209310080942603851;
+----------------------+----------------------------------------------------------------------------------------+---------+
| id                   | name                                                                                   | current |
+----------------------+----------------------------------------------------------------------------------------+---------+
| -5209310080942603851 | .chloeb-test-secret-20230307-2.deleted.1678222578.2586884d-adbc-4f6e-beff-be0d74d242fd |    NULL |
+----------------------+----------------------------------------------------------------------------------------+---------+
1 row in set (0.01 sec)

> select id, name, current from deleted_secrets where id=-5209310080942603851;
+----------------------+-------------------------------+----------------------+
| id                   | name                          | current              |
+----------------------+-------------------------------+----------------------+
| -5209310080942603851 | chloeb-test-secret-20230307-2 | -8339908504538408735 |
+----------------------+-------------------------------+----------------------+
1 row in set (0.00 sec)

> select id, groupid, secretid from accessgrants where secretid=-5209310080942603851;
Empty set (0.00 sec)

> select id, groupid, secretid from deleted_accessgrants where secretid=-5209310080942603851;
+----------+---------+----------------------+
| id       | groupid | secretid             |
+----------+---------+----------------------+
| 11036182 | 1297921 | -5209310080942603851 |
+----------+---------+----------------------+
1 row in set (0.00 sec)
```

Undelete the secret:

```
keywhiz.cli undelete secret --id -5209310080942603851
```

See if it looks undeleted now:

```
> keywhiz.cli describe secret --name chloeb-test-secret-20230307-2 --include-deleted
chloeb-test-secret-20230307-2
	Internal ID:
		-5209310080942603851
	Owner:
	Groups:
		chloeb-test-group
	Clients:
	Content HMAC:
		...
	Metadata:
	Created by:
		chloeb
	Created at:
		2023-03-07T20:54:50Z
	Updated by:
		chloeb
	Updated at:
		2023-03-28T21:35:42Z
	Content created by:
		chloeb
	Content created at:
		2023-03-07T20:54:50Z
Deleted Secrets: found 0
```

And see if it looks undeleted in the db:

```
> select id, name, current from secrets where id=-5209310080942603851;
+----------------------+-------------------------------+----------------------+
| id                   | name                          | current              |
+----------------------+-------------------------------+----------------------+
| -5209310080942603851 | chloeb-test-secret-20230307-2 | -8339908504538408735 |
+----------------------+-------------------------------+----------------------+
1 row in set (0.00 sec)

> select id, name, current from deleted_secrets where id=-5209310080942603851;
Empty set (0.00 sec)

> select id, groupid, secretid from accessgrants where secretid=-5209310080942603851;
+----------+---------+----------------------+
| id       | groupid | secretid             |
+----------+---------+----------------------+
| 11090385 | 1297921 | -5209310080942603851 |
+----------+---------+----------------------+
1 row in set (0.00 sec)

> select id, groupid, secretid from deleted_accessgrants where secretid=-5209310080942603851;
Empty set (0.00 sec)
```

Looks good!

## Test case 2

Let's make sure it "works" (doesn't work) if you try to undelete a secret where there is already another non-deleted secret with the same name:

```
> echo "foo" | keywhiz.cli add secret --name chloeb-test-secret-20230328
> echo "foo" | keywhiz.cli add secret --name chloeb-test-secret-20230328
Exception in thread "main" java.lang.AssertionError: Secret already exists.
	at keywhiz.cli.commands.AddAction.run(AddAction.java:88)
	at keywhiz.cli.CommandExecutor.executeCommand(CommandExecutor.java:171)
	at keywhiz.cli.CliMain.main(CliMain.java:63)
> keywhiz.cli assign secret --name chloeb-test-secret-20230328 --group chloeb-test-group
> keywhiz.cli delete secret --name chloeb-test-secret-20230328
Please confirm deletion of secret 'chloeb-test-secret-20230328': Y/N
Y
> keywhiz.cli describe secret --name chloeb-test-secret-20230328 --include-deleted
No non-deleted secret found.
Deleted Secrets: found 1
chloeb-test-secret-20230328
	Internal ID:
		-7573472852360510169
	Owner:
	Created by:
		chloeb
	Created at:
		2023-03-28T21:38:55Z
	Updated by:
		chloeb
	Updated at:
		2023-03-28T21:38:55Z
> echo "bar" | keywhiz.cli add secret --name chloeb-test-secret-20230328
> keywhiz.cli describe secret --name chloeb-test-secret-20230328 --include-deleted
chloeb-test-secret-20230328
	Internal ID:
		3278145263554138867
	Owner:
	Groups:
	Clients:
	Content HMAC:
		...
	Metadata:
	Created by:
		chloeb
	Created at:
		2023-03-28T21:40:03Z
	Updated by:
		chloeb
	Updated at:
		2023-03-28T21:40:03Z
	Content created by:
		chloeb
	Content created at:
		2023-03-28T21:40:03Z
Deleted Secrets: found 1
chloeb-test-secret-20230328
	Internal ID:
		-7573472852360510169
	Owner:
	Created by:
		chloeb
	Created at:
		2023-03-28T21:38:55Z
	Updated by:
		chloeb
	Updated at:
		2023-03-28T21:38:55Z
> keywhiz.cli undelete secret --id 3278145263554138867
2023-03-29 21:53:56,194 - Conflict
2023-03-29 21:53:56,196 - {"code":409,"message":"Cannot undelete secret since there is already a non-deleted secret with the same name"}
Exception in thread "main" java.lang.AssertionError: Conflict
	at keywhiz.cli.commands.UndeleteAction.undeleteSecret(UndeleteAction.java:67)
	at keywhiz.cli.commands.UndeleteAction.run(UndeleteAction.java:49)
	at keywhiz.cli.CommandExecutor.executeCommand(CommandExecutor.java:183)
	at keywhiz.cli.CliMain.main(CliMain.java:63)

> keywhiz.cli describe secret --name chloeb-test-secret-20230328 --include-deleted
chloeb-test-secret-20230328
	Internal ID:
		3278145263554138867
	Owner:
	Groups:
	Clients:
	Content HMAC:
		...
	Metadata:
	Created by:
		chloeb
	Created at:
		2023-03-28T21:40:03Z
	Updated by:
		chloeb
	Updated at:
		2023-03-28T21:40:03Z
	Content created by:
		chloeb
	Content created at:
		2023-03-28T21:40:03Z
Deleted Secrets: found 1
chloeb-test-secret-20230328
	Internal ID:
		-7573472852360510169
	Owner:
	Created by:
		chloeb
	Created at:
		2023-03-28T21:38:55Z
	Updated by:
		chloeb
	Updated at:
		2023-03-28T21:38:55Z
```

It correctly does not allow undeletion.

## Test case 3

Let's test undeleting an invalid secret ID:

```
> keywhiz.cli undelete secret --id -7573472852360
2023-03-29 21:52:15,489 - Not Found
2023-03-29 21:52:15,491 - {"code":404,"message":"No soft-deleted secret with the provided ID was found."}
Exception in thread "main" java.lang.AssertionError: Not Found
	at keywhiz.cli.commands.UndeleteAction.undeleteSecret(UndeleteAction.java:67)
	at keywhiz.cli.commands.UndeleteAction.run(UndeleteAction.java:49)
	at keywhiz.cli.CommandExecutor.executeCommand(CommandExecutor.java:183)
	at keywhiz.cli.CliMain.main(CliMain.java:63)
```

## Test case 4

Undeleting a "legacy" deleted secret (that doesn't exist in the deleted_secrets table):

```
> keywhiz.cli undelete secret --id -7573472852360510169
password for 'chloeb': 
2023-03-29 21:51:35,907 - Internal Server Error
2023-03-29 21:51:35,911 - Soft-deleted secret was found but does not have all of the data needed for undeletion. It may have been deleted before undeletion was supported.
Exception in thread "main" java.lang.AssertionError: Internal Server Error
	at keywhiz.cli.commands.UndeleteAction.undeleteSecret(UndeleteAction.java:67)
	at keywhiz.cli.commands.UndeleteAction.run(UndeleteAction.java:49)
	at keywhiz.cli.CommandExecutor.executeCommand(CommandExecutor.java:183)
	at keywhiz.cli.CliMain.main(CliMain.java:63)
```

### CODE REVIEW

File: cli/src/main/java/keywhiz/cli/configs/UndeleteActionConfig.java [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+import com.beust.jcommander.Parameter;
+import com.beust.jcommander.Parameters;
+
+@Parameters(commandDescription = "Undelete a secret that has previously been soft-deleted")
+public class UndeleteActionConfig {
+
+  @Parameter(description = "<secret>")
+  public String objectType;
+
+  @Parameter(names = "--id", description = "ID of the item to delete", required = true)
```
Comment (mmontgomery-square) [inline cli/src/main/java/keywhiz/cli/configs/UndeleteActionConfig.java:28]: Depending on whether this is truly secret-specific or not, we may want to be consistent in terminology between lines 22 and 28 (undelete secret/id of secret, etc.).

File: client/src/main/java/keywhiz/client/KeywhizClient.java [hunk 2]
```diff
@@ -274,6 +274,24 @@ public void deleteSecretWithId(long secretId) throws IOException {
...
+        .addPathSegment("undelete")
+        .addPathSegment(secretId.toString())
+        .build();
+
+    RequestBody body = RequestBody.create(JSON, mapper.writeValueAsString(null));
+    Request request = new Request.Builder()
+        .url(url)
+        .post(body)
+        .addHeader(HttpHeaders.CONTENT_TYPE, JSON.toString())
+        .build();
```
Comment (mmontgomery-square) [inline client/src/main/java/keywhiz/client/KeywhizClient.java:290]: Do we need to create a body and specify a content type even though there's no content? Does this have to do with how the requests are mapped/handled on the service side?

File: server/src/main/java/keywhiz/service/daos/SecretSeriesDAO.java [hunk 3]
```diff
@@ -514,6 +514,56 @@ private static void softDeleteSecretSeries(DSLContext dslContext, SecretsRecord
         .execute();
   }
 
+  public void undeleteSoftDeletedSecretSeriesByID(long id) {
+    dslContext.transaction(configuration -> {
+      undeleteSoftDeletedSecretSeriesByID(DSL.using(configuration), id);
+    });
+  }
+
+  private static void undeleteSoftDeletedSecretSeriesByID(
```
Comment (mmontgomery-square) [inline server/src/main/java/keywhiz/service/daos/SecretSeriesDAO.java:523]: Same for these two, ID in the name should be Id.

File: cli/src/main/java/keywhiz/cli/commands/UndeleteAction.java [hunk 4]
```diff
@@ -0,0 +1,73 @@
+/*
+ * Copyright (C) 2015 Square, Inc.
+ *
```
[the remark this answers is not in the document]
Comment (mmontgomery-square) [reply cli/src/main/java/keywhiz/cli/commands/UndeleteAction.java:3]: Just to be clear, I don't mean to do this now or as part of this PR, just in general.

File: cli/src/main/java/keywhiz/cli/CommandExecutor.java [hunk 5]
```diff
@@ -66,6 +68,7 @@ public enum Command {
     ADD,
     ASSIGN,
     DELETE,
+    UNDELETE,
     DESCRIBE,
     LIST,
     LOGIN,
```
[the remark this answers is not in the document]
Comment (mmontgomery-square) [reply cli/src/main/java/keywhiz/cli/CommandExecutor.java]: I tried fixing this by editing in the UI, hopefully that worked out.

**Review by mmontgomery-square — APPROVED** [review]
Just minor comments

File: cli/src/main/java/keywhiz/cli/CommandExecutor.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (chloebarker) [reply cli/src/main/java/keywhiz/cli/CommandExecutor.java]: oh good point

File: cli/src/main/java/keywhiz/cli/commands/UndeleteAction.java [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+    }
+
+    String trimmedType = type.toLowerCase().trim();
+    switch (trimmedType) {
+      case "secret":
+        undeleteSecret(id);
+        break;
+      default:
+        throw new IllegalArgumentException("Invalid undelete type specified: " + trimmedType);
+    }
```
[the remark this answers is not in the document]
Comment (chloebarker) [reply cli/src/main/java/keywhiz/cli/commands/UndeleteAction.java:53]: Not necessarily intending to eventually support undeletion of other types. This is more for consistency with the other Actions (e.g., DeleteAction.) I like this idea though

File: client/src/main/java/keywhiz/client/KeywhizClient.java (hunk 2 above)
Comment (chloebarker) [reply client/src/main/java/keywhiz/client/KeywhizClient.java:290]: I'm not sure. I borrowed this from another method that makes a POST request.

**Pull request merged**